### PR TITLE
Enable direct construction of compact Timestream Maps

### PR DESCRIPTION
**Pull request by cnweaver**

### CODE REVIEW

File: cmake/G3Test.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#ifndef G3TEST_H_INCLUDED
```
Comment (arahlin) [inline cmake/G3Test.h:1]: I think this header and associated cxx files should probably be in `core` so that other packages can use this test infrastructure.  Part of me also wants this infrastructure to be a separate PR altogether, so that it's abundantly clear that this is a new feature.

File: core/include/core/G3Timestream.h [hunk 2]
```diff
@@ -143,31 +144,107 @@ class G3Timestream : public G3FrameObject {
...
 	void Compactify();
+	
+	/// Construct a map whose underlying data storage is a contiguous 2D block.
+	/// \param keys the timestream keys for which the map should be constructed
+	/// \param start the start time which will be shared by all time streams
+	/// \param stop the stop time which will be shared by all time streams
+	/// \param n_samples the number of samples each timestream will have
+	/// \pre keys must be in sorted order
+	template<typename SampleType>
+	static G3TimestreamMap MakeCompact(const std::vector<std::string>& keys, G3Time start, G3Time stop, std::size_t n_samples){
```
Comment (arahlin) [inline core/include/core/G3Timestream.h:214]: Do we need / want python bindings for these methods?

**Review by arahlin — COMMENTED** [review]
Is possible to run these tests in python, or is the intention here for this MakeCompact infrastructure to be entirely in C++?  I suspect having access to this functionality in python would be useful in general.

Also I think the C++ test infrastructure should probably be a separate PR for clarity.

**Comment by nwhitehorn** [comment]
With respect to the first point, this interface is already exposed to Python, but differently since it involves numpy. This brings C++ up to parity with Python, but it needs different tests since the API has to be different. Hence they need to be C++ tests.

File: cmake/G3Test.h (hunk 1 above)
Comment (cnweaver) [reply cmake/G3Test.h:1]: My logic for putting it in `cmake` was that exactly so that it should be accessible to all packages, and `cmake` seemed more fundamental than even `core`. (Also, IceCube happens to keep this sort of thing in `cmake`.) However, given that all other packages depend on `core`, I think it could go there just as well. 

If you would like this as a separate PR, that would be doable; I think the main reason I didn't try to split it was that I didn't have any other immediate uses, so it seemed like a PR for the test stuff alone would not demonstrate that it really worked. I could probably invent such a minimal use, though.

File: core/include/core/G3Timestream.h (hunk 2 above)
Comment (cnweaver) [reply core/include/core/G3Timestream.h:214]: As Nathan has mentioned, the idea in large part was to enable doing what could already be done from python (via `timestream_from_iterable`, I think), but directly in C++ without numpy. As such, I think python mostly doesn't need this. However, it looks like the features may not be entirely synchronized, as it looks like python can set the units as part of the same operation, while I included start and stop time instead. I could be useful to make both interfaces support all of those.

File: cmake/G3Test.h (hunk 1 above)
Comment (arahlin) [reply cmake/G3Test.h:1]: Given the way that spt3g_software is used by SO (i.e. as a separate dependency), it would be nice to allow building a test program by linking against the core library and supplying the right header.

File: core/include/core/G3Timestream.h (hunk 2 above)
Comment (arahlin) [reply core/include/core/G3Timestream.h:214]: Agreed on that last point -- both interfaces should support any of the configurable options, including compression level.

File: cmake/G3Test.h (hunk 1 above)
Comment (cnweaver) [reply cmake/G3Test.h:1]: Yes, that is a good point. I forgot that headers for `cmake` are not installed. Unfortunately, the linking aspect is a little tricky, since the test program needs to have a `main()`, while it would be rather inappropriate for that to appear as a symbol in `libspt3g-core.so`. I'll do some brainstorming to figure out how a downstream project could utilize this. Maybe some shenanigans with hiding the actual implementation code in a macro in the header, and some kind of cmake logic to autogenerate an implmentation file into which that is expanded?

File: core/include/core/G3Timestream.h (hunk 2 above)
Comment (cnweaver) [reply core/include/core/G3Timestream.h:214]: Looking more carefully, I think the corresponding function is actually `G3TimestreamMap_from_numpy()`, which has `keys`, `data`, `start`, `stop`, `units`, and `copy_data`. I can add `units` to `MakeCompact` and, `compression_level` to both. Does it seem worthwhile to implement `copy_data` for `MakeCompact`? I can do it, but I personally don't ever want to use it, and I'm not sure if anyone else would either.

Comment (arahlin) [reply core/include/core/G3Timestream.h:214]: I'm fine with not implementing copy_data in C++!

File: cmake/Spt3gIncludes.cmake.in [hunk 3]
```diff
@@ -74,3 +74,39 @@ macro(add_spt3g_test test_name)
...
+	                      "SOURCE_FILES;TEST_LABELS;USE_PROJECTS" # multi-value arguments
+	                      ${ARGN}
+	                      )
+	
+	add_test(NAME ${PROJECT}/${test_name}
+	         WORKING_DIRECTORY ${SPT3G_BUILD_DIR}
+	         COMMAND ${PROJECT}-${test_name}
+	         )
+	
+	if(ADD_TEST_PROGRAM_TEST_LABELS)
```
Comment (arahlin) [inline cmake/Spt3gIncludes.cmake.in:91]: Something weird going on with the indentation in this code block.  Mixing tabs and spaces maybe?

Comment (cnweaver) [reply cmake/Spt3gIncludes.cmake.in:91]: That was intentional, as the tabs provide the structural indentation, and the spaces align things within the lines, but I guess it's a dying style these days. I can make it use only one character is that is strongly preferred.

Comment (arahlin) [reply cmake/Spt3gIncludes.cmake.in:91]: I guess the github web interface just formats it oddly?  Some of the lines don't look like they're aligned like you'd expect.

Comment (cnweaver) [reply cmake/Spt3gIncludes.cmake.in:91]: Curious; it seems to work as expected for me even when I adjust the displayed tab width:
4 space tab stops:
Screen Shot 2023-04-07 at 15 36 49
8 space tab stops:
Screen Shot 2023-04-07 at 15 36 19

Comment (arahlin) [reply cmake/Spt3gIncludes.cmake.in:91]: Oh!  It's the blocks below that, lines 91-99.

File: cmake/G3Test.h (hunk 1 above)
Comment (cnweaver) [reply cmake/G3Test.h:1]: This is now done; auto-generating an implementation file in the build directory is very slightly evil, but pretty simple.

File: core/include/core/G3Timestream.h (hunk 2 above)
Comment (cnweaver) [reply core/include/core/G3Timestream.h]: This should now be cleaned up.

File: cmake/Spt3gIncludes.cmake.in (hunk 3 above)
Comment (cnweaver) [reply cmake/Spt3gIncludes.cmake.in:101]: Thanks, there was a mistake there, which is hopefully now fixed.

**Review by arahlin — APPROVED** [review]
